File: backend/app/schemas/faculty.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional, Any
from datetime import date, datetime


# Faculty Profile
class FacultyProfileResponse(BaseModel):
    id: int
    user_id: int
    employee_id: str
    department: str
    designation: str
    office_room: str
    assigned_subjects: str
    assigned_sections: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Schedule Item
class FacultyScheduleItem(BaseModel):
    id: int
    day_of_week: str
    period_number: int
    start_time: str
    end_time: str
    subject_name: str
    subject_code: str
    section: str
    classroom: str
    model_config = ConfigDict(from_attributes=True)


# Stats & Dashboard Overview
class FacultyDashboardStats(BaseModel):
    today_classes_count: int
    total_assigned_students: int
    pending_submissions_count: int
    average_class_attendance: float
    total_quizzes_created: int
    question_papers_uploaded: int


class FacultyDashboardResponse(BaseModel):
    profile: FacultyProfileResponse
    stats: FacultyDashboardStats
    today_schedule: List[FacultyScheduleItem]
    upcoming_classes: List[FacultyScheduleItem]
    assigned_subjects: List[str]
    assigned_sections: List[str]
    notifications: List[dict]


# Attendance
class AttendanceMarkInput(BaseModel):
    student_id: Optional[int] = None
    student_name: str
    register_number: str
    subject_code: str
    section: str
    date: date
    status: str  # Present, Absent, Late, Leave
    remarks: Optional[str] = None


class BulkAttendanceInput(BaseModel):
    subject_code: str
    section: str
    date: date
    records: List[AttendanceMarkInput]


class AttendanceRecordResponse(BaseModel):
    id: int
    student_id: Optional[int]
    student_name: str
    register_number: str
    subject_code: str
    section: str
    date: date
    status: str
    remarks: Optional[str]
    model_config = ConfigDict(from_attributes=True)


# Assignments
class FacultyAssignmentInput(BaseModel):
    title: str
    subject_code: str
    section: str
    description: Optional[str] = None
    due_date: str
    max_marks: int = 100
    attachment_url: Optional[str] = None


class FacultyAssignmentResponse(BaseModel):
    id: int
    faculty_id: int
    title: str
    subject_code: str
    section: str
    description: Optional[str]
    due_date: str
    max_marks: int
    attachment_url: Optional[str]
    submissions_count: int = 0
    graded_count: int = 0
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class GradeSubmissionInput(BaseModel):
    grade: str
    remarks: Optional[str] = None


class FacultySubmissionResponse(BaseModel):
    id: int
    assignment_id: int
    student_id: Optional[int]
    student_name: str
    register_number: str
    submission_text: Optional[str]
    file_url: Optional[str]
    submitted_at: datetime
    grade: Optional[str]
    remarks: Optional[str]
    status: str
    model_config = ConfigDict(from_attributes=True)


# Question Papers
class FacultyQuestionPaperInput(BaseModel):
    title: str
    subject_name: str
    subject_code: str
    department: str
    semester: int
    academic_year: str
    exam_type: str
    pdf_url: Optional[str] = None


class FacultyQuestionPaperResponse(BaseModel):
    id: int
    faculty_id: int
    title: str
    subject_name: str
    subject_code: str
    department: str
    semester: int
    academic_year: str
    exam_type: str
    pdf_url: Optional[str]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Quizzes
class FacultyQuizInput(BaseModel):
    title: str
    subject_code: str
    section: str
    num_questions: int = 5
    duration_minutes: int = 15
    total_marks: int = 50
    questions_json: Optional[Any] = None


class FacultyQuizResponse(BaseModel):
    id: int
    faculty_id: int
    title: str
    subject_code: str
    section: str
    num_questions: int
    duration_minutes: int
    total_marks: int
    is_published: bool
    questions_json: Optional[Any]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class FacultyQuizResultResponse(BaseModel):
    id: int
    quiz_id: int
    student_id: Optional[int]
    student_name: str
    score: int
    total_marks: int
    submitted_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Timetable
class FacultyTimetableRequestInput(BaseModel):
    request_date: date
    current_period: int
    requested_period: int
    reason: str


class FacultyTimetableRequestResponse(BaseModel):
    id: int
    faculty_id: int
    request_date: date
    current_period: int
    requested_period: int
    reason: str
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# Student Roster
class StudentRosterItem(BaseModel):
    id: int
    student_name: str
    register_number: str
    department: str
    semester: int
    section: str
    attendance_percentage: float
    assignment_status: str  # All Completed, Pending, Overdue

File: frontend/src/api/faculty.ts
import axios from 'axios';

const API_BASE = (import.meta.env.VITE_API_URL || 'http://localhost:8000') + '/api/v1/faculty';

axios.interceptors.request.use((config) => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

const getAuthHeader = () => {
  const token = localStorage.getItem('token');
  return token ? { headers: { Authorization: `Bearer ${token}` } } : {};
};


export interface FacultyProfile {
  id: number;
  user_id: number;
  employee_id: string;
  department: string;
  designation: string;
  office_room: string;
  assigned_subjects: string;
  assigned_sections: string;
  created_at: string;
}

export interface FacultyScheduleItem {
  id: number;
  day_of_week: string;
  period_number: number;
  start_time: string;
  end_time: string;
  subject_name: string;
  subject_code: string;
  section: string;
  classroom: string;
}

export interface FacultyDashboardStats {
  today_classes_count: number;
  total_assigned_students: number;
  pending_submissions_count: number;
  average_class_attendance: number;
  total_quizzes_created: number;
  question_papers_uploaded: number;
}

export interface FacultyDashboardData {
  profile: FacultyProfile;
  stats: FacultyDashboardStats;
  today_schedule: FacultyScheduleItem[];
  upcoming_classes: FacultyScheduleItem[];
  assigned_subjects: string[];
  assigned_sections: string[];
  notifications: Array<{
    id: number;
    title: string;
    message: string;
    time: string;
    type: string;
  }>;
}

export interface AttendanceRecord {
  id: number;
  student_id?: number;
  student_name: string;
  register_number: string;
  subject_code: string;
  section: string;
  date: string;
  status: string;
  remarks?: string;
}

export interface FacultyAssignment {
  id: number;
  faculty_id: number;
  title: string;
  subject_code: string;
  section: string;
  description?: string;
  due_date: string;
  max_marks: number;
  attachment_url?: string;
  submissions_count: number;
  graded_count: number;
  created_at: string;
}

export interface FacultySubmission {
  id: number;
  assignment_id: number;
  student_id?: number;
  student_name: string;
  register_number: string;
  submission_text?: string;
  file_url?: string;
  submitted_at: string;
  grade?: string;
  remarks?: string;
  status: string;
}

export interface FacultyQuestionPaper {
  id: number;
  faculty_id: number;
  title: string;
  subject_name: string;
  subject_code: string;
  department: string;
  semester: number;
  academic_year: string;
  exam_type: string;
  pdf_url?: string;
  created_at: string;
}

export interface FacultyQuiz {
  id: number;
  faculty_id: number;
  title: string;
  subject_code: string;
  section: string;
  num_questions: number;
  duration_minutes: number;
  total_marks: number;
  is_published: boolean;
  questions_json?: any;
  created_at: string;
}

export interface FacultyQuizResult {
  id: number;
  quiz_id: number;
  student_id?: number;
  student_name: string;
  score: number;
  total_marks: number;
  submitted_at: string;
}

export interface StudentRosterItem {
  id: number;
  student_name: string;
  register_number: string;
  department: string;
  semester: number;
  section: string;
  attendance_percentage: number;
  assignment_status: string;
}

export const facultyApi = {
  getDashboard: async (): Promise<FacultyDashboardData> => {
    const res = await axios.get(`${API_BASE}/dashboard`);
    return res.data;
  },

  getAttendance: async (subject_code?: string, section?: string, attendance_date?: string): Promise<AttendanceRecord[]> => {
    const res = await axios.get(`${API_BASE}/attendance`, { params: { subject_code, section, attendance_date } });
    return res.data;
  },

  markAttendance: async (data: Partial<AttendanceRecord>): Promise<AttendanceRecord> => {
    const res = await axios.post(`${API_BASE}/attendance`, data);
    return res.data;
  },

  markBulkAttendance: async (subject_code: string, section: string, date: string, records: Partial<AttendanceRecord>[]): Promise<AttendanceRecord[]> => {
    const res = await axios.post(`${API_BASE}/attendance/bulk`, { subject_code, section, date, records });
    return res.data;
  },

  editAttendance: async (id: number, statusVal: string, remarks?: string): Promise<AttendanceRecord> => {
    const res = await axios.put(`${API_BASE}/attendance/${id}`, null, { params: { status_val: statusVal, remarks } });
    return res.data;
  },

  getAssignments: async (): Promise<FacultyAssignment[]> => {
    const res = await axios.get(`${API_BASE}/assignments`);
    return res.data;
  },

  createAssignment: async (data: Partial<FacultyAssignment>): Promise<FacultyAssignment> => {
    const res = await axios.post(`${API_BASE}/assignments`, data);
    return res.data;
  },

  editAssignment: async (id: number, data: Partial<FacultyAssignment>): Promise<FacultyAssignment> => {
    const res = await axios.put(`${API_BASE}/assignments/${id}`, data);
    return res.data;
  },

  deleteAssignment: async (id: number): Promise<void> => {
    await axios.delete(`${API_BASE}/assignments/${id}`);
  },

  getSubmissions: async (assignmentId: number): Promise<FacultySubmission[]> => {
    const res = await axios.get(`${API_BASE}/assignments/${assignmentId}/submissions`);
    return res.data;
  },

  gradeSubmission: async (submissionId: number, grade: string, remarks?: string): Promise<FacultySubmission> => {
    const res = await axios.post(`${API_BASE}/submissions/${submissionId}/grade`, { grade, remarks });
    return res.data;
  },

  getQuestionPapers: async (): Promise<FacultyQuestionPaper[]> => {
    const res = await axios.get(`${API_BASE}/question-papers`);
    return res.data;
  },

  uploadQuestionPaper: async (data: Partial<FacultyQuestionPaper>): Promise<FacultyQuestionPaper> => {
    const res = await axios.post(`${API_BASE}/question-papers`, data);
    return res.data;
  },

  deleteQuestionPaper: async (id: number): Promise<void> => {
    await axios.delete(`${API_BASE}/question-papers/${id}`);
  },

  getQuizzes: async (): Promise<FacultyQuiz[]> => {
    const res = await axios.get(`${API_BASE}/quizzes`);
    return res.data;
  },

  createQuiz: async (data: Partial<FacultyQuiz>): Promise<FacultyQuiz> => {
    const res = await axios.post(`${API_BASE}/quizzes`, data);
    return res.data;
  },

  togglePublishQuiz: async (id: number): Promise<FacultyQuiz> => {
    const res = await axios.put(`${API_BASE}/quizzes/${id}/publish`);
    return res.data;
  },

  deleteQuiz: async (id: number): Promise<void> => {
    await axios.delete(`${API_BASE}/quizzes/${id}`);
  },

  getQuizScores: async (quizId: number): Promise<FacultyQuizResult[]> => {
    const res = await axios.get(`${API_BASE}/quizzes/${quizId}/scores`);
    return res.data;
  },

  getTimetable: async (): Promise<FacultyScheduleItem[]> => {
    const res = await axios.get(`${API_BASE}/timetable`);
    return res.data;
  },

  requestTimetableChange: async (data: { request_date: string; current_period: number; requested_period: number; reason: string }) => {
    const res = await axios.post(`${API_BASE}/timetable/change-request`, data);
    return res.data;
  },

  getStudents: async (params?: { department?: string; semester?: number; section?: string; search?: string }): Promise<StudentRosterItem[]> => {
    const res = await axios.get(`${API_BASE}/students`, { params });
    return res.data;
  },
};
